org.eclipse.xtext/src/org/eclipse/xtext/validation/DefaultUniqueNameContext.java

```diff
@@ -0,0 +1,320 @@
+
+	/**
+	 * Base class for {@link INamesAreUniqueValidationHelper.ContextProvider}.
+	 */
+	public static abstract class BaseContextProvider implements INamesAreUniqueValidationHelper.ContextProvider {
+		@Inject
+		private IResourceServiceProvider.Registry resourceServiceProviderRegistry = IResourceServiceProvider.Registry.INSTANCE;
+		@Inject
+		private ICaseInsensitivityHelper caseInsensitivityHelper = new CaseInsensitivityHelper();
+
+		protected IResourceServiceProvider getResourceServiceProvider(Resource r) {
+			return resourceServiceProviderRegistry.getResourceServiceProvider(r.getURI());
+		}
+
+		protected IResourceDescription getResourceDescription(Resource resource) {
+			IResourceDescription.Manager manager = getResourceDescriptionManager(resource);
+			if (manager != null) {
+				return manager.getResourceDescription(resource);
+			}
+			return null;
+		}
+
+		protected IResourceDescription.Manager getResourceDescriptionManager(Resource resource) {
+			IResourceServiceProvider resourceServiceProvider = getResourceServiceProvider(resource);
+			if (resourceServiceProvider == null)
+				return null;
+			return resourceServiceProvider.getResourceDescriptionManager();
+		}
+
+		protected ICaseInsensitivityHelper getCaseInsensitivityHelper() {
+			return caseInsensitivityHelper;
+		}
+
+		protected void setResourceServiceProviderRegistry(
+				IResourceServiceProvider.Registry resourceServiceProviderRegistry) {
+			this.resourceServiceProviderRegistry = resourceServiceProviderRegistry;
+		}
+	}
+
+	/**
+	 * Base class for non-local unique name validation context providers.
+	 */
+	public static abstract class BaseGlobalContextProvider extends BaseContextProvider {
+		@Inject
+		private IResourceDescriptionsProvider indexAccess;
+
+		protected IResourceDescriptions getIndex(Resource resource) {
+			ResourceSet resourceSet = resource.getResourceSet();
+			if (resourceSet == null) {
+				return null;
+			}
+			return indexAccess.getResourceDescriptions(resourceSet);
+		}
+
+		protected boolean intersects(IResourceDescription left, IResourceDescription right, boolean caseSensitive) {
+			for (IEObjectDescription description : left.getExportedObjects()) {
+				Iterable<IEObjectDescription> exportedObjects = right.getExportedObjects(EcorePackage.Literals.EOBJECT,
+						description.getName(), !caseSensitive);
+				if (!Iterables.isEmpty(exportedObjects)) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		protected boolean isAffected(IResourceDescription.Delta delta, IResourceDescription candidate,
+				boolean caseSensitive) {
+			if (candidate.getURI().equals(delta.getUri())) {
+				return false;
+			}
+			if (delta.getNew() != null && intersects(delta.getNew(), candidate, caseSensitive)) {
+				return true;
+			}
+			if (delta.getOld() != null && intersects(delta.getOld(), candidate, caseSensitive)) {
+				return true;
+			}
+			return false;
+		}
+	}
+
+	/**
+	 * <p>
+	 * Provide a context for the entire index.
+	 * </p></p>
```

Comment:
double closing
BTW: I placed the `<p>` remarks on the second paragraphs of the javadocs. The first paragraph does not need to be encapsulated, but it does not hurt and I'm fine with that.